packages/plugin-assets/src/upload.js

```diff
@@ -0,0 +1,253 @@
+    assetVersion,
+    assetVersions = [];
+
+  const debug = message => {
+    const wasSpinning = spinner.isSpinning;
+    spinner.stop();
+    logger.debug(message);
+    if (wasSpinning) {
+      spinner.start();
+    }
+  };
+
+  const handleError = (message, error) => {
+    spinner.stop();
+    if (error) {
+      debug(error.toString());
+    }
+    throw new TwilioCliError(message);
+  };
+
+  spinner.start('Loading config');
+  const config = await pluginConfig.getConfig();
+  if (
+    config[accountSid] &&
+    config[accountSid].serviceSid &&
+    config[accountSid].environmentSid
+  ) {
+    const { serviceSid, environmentSid } = config[accountSid];
+    const client = new TwilioServerlessApiClient({
+      username: apiKey,
+      password: apiSecret,
+    });
+    spinner.text = 'Checking assets service';
+    try {
+      debug(
+        `Fetching environment with sid ${environmentSid} from service with sid ${serviceSid}`
+      );
+      environment = await getEnvironment(environmentSid, serviceSid, client);
+    } catch (error) {
+      handleError('Could not fetch asset service environment', error);
+    }
+    try {
+      debug(`Checking for functions in service with sid ${serviceSid}`);
+      const functions = await listFunctionResources(serviceSid, client);
+      if (functions.length > 0) {
+        handleError('Not an Asset Plugin service: service contains functions');
+      }
+    } catch (error) {
+      handleError(
+        'Could not fetch last build of asset service environment',
+        error
+      );
+    }
+    spinner.text = 'Preparing asset';
+    try {
+      filePath = resolve(file);
+      debug(`Reading file from ${filePath}`);
+      assetContent = await readFile(filePath);
+    } catch (error) {
+      handleError(`Could not read ${filePath}`, error);
+    }
+    const path = `/${basename(filePath)}`;
+    const newAsset = {
+      name: path,
+      access: 'public',
+      path,
+      content: assetContent,
+    };
+
+    try {
+      debug('Finding existing asset resources');
+      existingAssets = await listAssetResources(serviceSid, client);
+    } catch (error) {
+      handleError(
+        `Could not load existing asset resources from service ${serviceSid}`,
+        error
+      );
+    }
+    const existingAsset = existingAssets.find(
+      asset => asset.friendly_name === newAsset.name
+    );
+    if (existingAsset) {
+      spinner.stop();
+      const answers = await inquirer.prompt([
+        {
+          message: `There is already an asset called ${existingAsset.friendly_name}. What do you want to do?`,
+          type: 'list',
+          name: 'conflict',
+          choices: ['Overwrite', 'Rename', 'Abort'],
+        },
+      ]);
+      if (answers.conflict === 'Overwrite') {
+        newAsset.sid = existingAsset.sid;
+        spinner.start('Overwriting existing asset');
+      } else if (answers.conflict === 'Rename') {
+        let newNameAnswers = await inquirer.prompt([
```

Comment:
We should do verification here that it doesn't contain invalid characters

Reply:
[Invalid characters are](https://www.twilio.com/docs/runtime/functions-assets-api/api/asset-version#create-an-asset-version-resource) `;,?:@+&$()' "` and fragments, so `#`.

Most of these can actually be uri encoded and still work fine. Rather than confuse the code (we should also check filenames which are passed in as the original file, for example) for an edge case I would prefer for now to fail the asset version upload with a better error message (I've added the asset path to the failure message).

We can add this sort of checking in later, if required.